socketEvent.js

```diff
@@ -0,0 +1,189 @@
+const {
+  request,
+  io,
+  raceRooms,
+  syncraceget,
+  paraTypedByUser,
+  expiryInSeconds,
+  TFSBackendURL,
+} = require("./utils.js");
+
+exports.disconnect = function (socket) {
+  socket.on("disconnect", () => {
+    raceRooms.keys("*", (err, rooms) => {
+      for (let i = 0; i < rooms.length; i++) {
+        raceRooms.get(rooms[i], (err, userintheroom) => {
+          json_data = JSON.parse(userintheroom);
+          for (let j = 0; j < json_data.length; j++) {
+            if (socket.id == json_data[j].socketId) {
+              io.to(rooms[i]).emit(
+                "DISCONNECTED",
+                `${json_data[j].name} has left the race`
+              );
+              json_data.splice(j, 1);
+              raceRooms.set(rooms[i], JSON.stringify(json_data));
+            }
+          }
+        });
+      }
+    });
+
+    console.log("Disconnected");
+  });
+};
+
+exports.createorjoinroom = function (socket) {
+  socket.on("create/join", async ({ room, username }) => {
+    /*  
+            {
+              'room1':[189351,{'user1:'name'},{'user2':'name'}],
+              'room2':{'user3':'name','user4':'name'},
+            } 
+        */
+    userExistsAlready = io.sockets.adapter.sids[socket.id][room];
+    if (userExistsAlready) {
+      io.to(socket.id).emit(
+        "ALREADY_JOINED",
+        "You have already joined the room!"
+      );
+    } else {
+      return syncraceget(room)
+        .then((res) => {
+          if (res != null) {
+            res = JSON.parse(res);
+            if (res[0] != true) {
+              userDetails = {};
+              userDetails.name = username;
+              userDetails.socketId = socket.id;
+              res.push(userDetails);
+              raceRooms.set(room, JSON.stringify(res));
+              return;
+            } else {
+              io.to(socket.id).emit("RACE_STARTED", "Race has already started");
+              return false;
+            }
+          } else {
+            time = Math.floor(Date.now() / 1000);
+            userDetails = {};
+            userDetails.name = username;
+            userDetails.socketId = socket.id;
+            raceRooms.set(
+              room,
+              JSON.stringify([time, userDetails]),
+              "EX",
+              expiryInSeconds
+            );
+            return;
+          }
+        })
+        .then((userAddedInRedis = true) => {
+          if (userAddedInRedis == true) {
+            socket.join(room);
+
+            // To get the clients of the room stored in the socket
+            // io.of("/")
+            //   .in(room)
+            //   .clients((err, client) => {
+            //     console.log(client);
+            //   });
+
+            raceRooms.get(room, (err, res) => {
+              res = JSON.parse(res);
+              usernames = res
+                .map((userData) => {
+                  if (userData.name != undefined) {
+                    return userData.name;
+                  }
```

Comment:
This is kinda confusing to read, we will get on a call for this.

Reply:
Cool